File: src/main/java/org/gga/graph/maps/DataGraph.java
package org.gga.graph.maps;

import com.google.common.base.Optional;
import org.gga.graph.Edge;
import org.gga.graph.Graph;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * @author mike
 */
public interface DataGraph<N, E> {
    int V();

    boolean isDirected();

    int getIndex(@Nonnull N data);

    @Nonnull
    N getNode(int v);

    @Nonnull
    Optional<N> getNodeSafe(int v);

    void setNode(int v, @Nonnull N data);

    @Nullable
    E edge(@Nonnull N n1, @Nonnull N n2);

    @Nullable
    E edge(int v1, int v2);

    @Nullable
    E getEdge(Edge e);

    @Nonnull
    Edge insert(@Nonnull N n1, @Nonnull N n2, @Nullable E e);

    @Nonnull
    Edge insert(int v1, int v2, @Nullable E e);

    void remove(Edge edge);

    Graph getIntGraph();

    Class<N> getNodeClass();
    Class<E> getEdgeClass();
}

File: src/main/scala/org/gga/graph/maps/DataGraph.scala
package org.gga.graph.maps

import org.gga.graph.Edge
import org.gga.graph.Graph

/**
 * @author mike
 */
trait DataGraph[N, E] {
  def V: Int
  def vertices: Range = Range(0, V)

  def isDirected: Boolean

  def getIndex(data: N): Option[Int]

  def node(v: Int): N

  def setNode(v: Int, data: N)

  def edge(n1: N, n2: N): Option[E]

  def edge(v1: Int, v2: Int): Option[E]

  def edge(e: Edge): E

  def insert(n1: N, n2: N, e: E): Option[Edge]

  def insert(v1: Int, v2: Int, e: E): Edge

  def remove(edge: Edge)

  def intGraph: Graph
}